src/backend/distributed/master/master_create_shards.c

```diff
@@ -217,8 +214,104 @@ master_create_worker_shards(PG_FUNCTION_ARGS)
+ */
+void
+CreateColocatedShards(Oid targetRelationId, Oid sourceRelationId)
+{
+	char *targetTableRelationOwner = NULL;
+	char targetShardStorageType = 0;
+	List *existingShardList = NIL;
+	List *sourceShardIntervalList = NIL;
+	List *targetTableDDLEvents = NIL;
+	ListCell *sourceShardCell = NULL;
+
+	/* make sure that tables are hash partitioned */
+	CheckHashPartitionedTable(targetRelationId);
+	CheckHashPartitionedTable(sourceRelationId);
+
+	/*
+	 * In contrast to append/range partitioned tables it makes more sense to
+	 * require ownership privileges - shards for hash-partitioned tables are
+	 * only created once, not continually during ingest as for the other
+	 * partitioning types.
+	 */
+	EnsureTableOwner(targetRelationId);
+
+	/* we plan to add shards: get an exclusive metadata lock on the target relation */
+	LockRelationDistributionMetadata(targetRelationId, ExclusiveLock);
+
+	/* a share metadata lock is enough on the source relation */
+	LockRelationDistributionMetadata(sourceRelationId, ShareLock);
+
+	/* prevent placement changes of the source relation until we colocate them */
+	sourceShardIntervalList = LoadShardIntervalList(sourceRelationId);
+	LockShards(sourceShardIntervalList, ShareLock);
+
+	/* validate that shards haven't already been created for this table */
+	existingShardList = LoadShardList(targetRelationId);
+	if (existingShardList != NIL)
+	{
+		char *targetRelationName = get_rel_name(targetRelationId);
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("table \"%s\" has already had shards created for it",
+							   targetRelationName)));
+	}
+
+	targetTableRelationOwner = TableOwner(targetRelationId);
+	targetTableDDLEvents = GetTableDDLEvents(targetRelationId);
+	targetShardStorageType = ShardStorageType(targetRelationId);
+
+	foreach(sourceShardCell, sourceShardIntervalList)
+	{
+		ShardInterval *sourceShardInterval = (ShardInterval *) lfirst(sourceShardCell);
+		uint64 sourceShardId = sourceShardInterval->shardId;
+		uint64 newShardId = GetNextShardId();
+		ListCell *sourceShardPlacementCell = NULL;
+
+		int32 shardMinValue = DatumGetInt32(sourceShardInterval->minValue);
+		int32 shardMaxValue = DatumGetInt32(sourceShardInterval->maxValue);
+		text *shardMinValueText = IntegerToText(shardMinValue);
+		text *shardMaxValueText = IntegerToText(shardMaxValue);
+
+		List *sourceShardPlacementList = ShardPlacementList(sourceShardId);
+		foreach(sourceShardPlacementCell, sourceShardPlacementList)
+		{
+			ShardPlacement *sourcePlacement =
+				(ShardPlacement *) lfirst(sourceShardPlacementCell);
+			char *sourceNodeName = sourcePlacement->nodeName;
+			int32 sourceNodePort = sourcePlacement->nodePort;
+
+			bool created = WorkerCreateShard(targetRelationId, sourceNodeName,
+											 sourceNodePort, newShardId,
+											 targetTableRelationOwner,
+											 targetTableDDLEvents);
+			if (created)
+			{
+				const RelayFileState shardState = FILE_FINALIZED;
```

Comment:
Once we fix #810 , we need to use the source placement's shard state instead of hard coded `FILE_FINALIZED` state. But, for now lets keep it as it is.